### ISSUE TITLE
Nullability control for Avro-formatted sinks

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#21722]
*Part of epic #17209.*

## Problem

Materialize cannot always infer whether an expression is non-nullable. However, users often want to coerce Materialize into treating as expression as non-nullable, e.g. to ensure that the Avro schema generated for a Kafka sink marks the field as non-nullable.

Here's an example. Suppose you have a `FULL OUTER JOIN` and coalesce the join key columns on both sides of the join, as in:

```sql
create table t1 (a int not null, b int not null);
create table t2 (b int not null, c int not null);
select coalesce(t1.b, t2.b) from t1 full outer join t2 on t1.b = t2.b;
```

The expression `coalesce(t1.b, t2.b)` will never be null, because the semantics of a `FULL OUTER JOIN` guarantee that at most one of `t1.b` or `t2.b` is null. But Materialize cannot infer that (it's relatively complicated to do so), and so it conservatively reports the type of that expression as possibly nullable.

### Option 1: Expose `mz_internal.error_if_null`

One way out is to allow our users to call the `mz_internal.error_if_null` function. Imagine:

```sql
select error_if_null(coalesce(t1.b, t2.b), 'FULL OUTER JOIN unexpectedly returned NULL') from t1 full outer join t2 on t1.b = t2.b;
```

The semantics are: if the first argument is null, produce an error with the message specified as the second argument; otherwise, if the first argument is non-null, return the first argument directly.

While we might be able to automatically infer the non-nullability in this specific example, the `error_if_null` function would be a general purpose tool that users could deploy in other situations where Materialize failed to infer non-nullability.

#### Open questions

  * Is `error_if_null` the right name?
  * Is there prior art for this function in other databases? 
  * Should users be able to choose the error code the function returns?
  * Can we solve the existing optimizer issues with `error_if_null`? https://github.com/MaterializeInc/materialize/issues/20199

### Option 2: `FORMAT AVRO` option

We could alternatively introduce an option like `FORCE NOT NULL ON` in the `FORMAT AVRO` clause of the `CREATE SINK` statement:

```sql
CREATE SINK ... INTO KAFKA
FORMAT AVRO (
    FORCE NOT NULL ON <field-specifier>
)
```

The `<field-specifier>` would work as it does for `DEFAULT ON` (https://github.com/MaterializeInc/materialize/issues/21662). When specified, the targeted field should will be generated with a non-nullable type in the Avro schema. If a `NULL` value presents, the sink emits an error ... somehow (see https://github.com/MaterializeInc/materialize/issues/17668).

### Evaluation

Philosophically, I much prefer option 1. Whether a field is known to be non-nullable depends of the definition of the view on which the sink is built. I think it's best to keep any non-nullability hints as close as possible to the logic that determines the non-nullability.  If the non-nullability hint is in the view itself, it's more likely you'll notice the hint if you're changing the logic in a way that invalidates the non-nullability; if the non-nullability hint is in a downstream sink, it's likely to go unnoticed.

That said, there are two points in favor of option 2:

  1. One user has specifically requested that these hints live on the sink rather than in the underlying view.
  2. @ggevay reports on Slack (see link below) that the `error_if_null` semantics are poorly defined and even the current behavior (it "mostly works") is hard to maintain through the optimizer. Option 2 sidesteps these issues, as the nullability hint is outside the optimizer's purview.

## See also

  * https://github.com/MaterializeInc/materialize/issues/22269
  * [Slack conversation](https://materializeinc.slack.com/archives/C02FWJ94HME/p1694555534517729)

**Comment by umanwizard** [comment]
My problem with approach 1 philosophically is that fundamentally, nullability is not part of the type in SQL, it's a constraint that the optimizer may or may not infer. On the other hand, nullability _is_ part of the type in Avro, and so it makes sense in my mind to express this at the point where we describe how to generate Avro schemas from Materialize schemas, which is exactly the `FORMAT` statement.

I don't really think we should ever be relying on the optimizer to figure out that a field is nullable. What we might want to do instead is define some syntactic properties that guarantee nullability -- perhaps one such property could indeed be "the top-level node in the `MirScalarExpr` is `mz_error_if_null`".

**Comment by umanwizard** [comment]
An idea @frankmcsherry proposed is that we _only_ mark fields as non-nullable in Avro schemas if non-nullability can be inferred from one of these stable syntactic properties that we define, and mark all other fields as nullable, even if today the optimizer is able to infer that some of them are non-nullable (because we don't consider the set of constraints the optimizer can infer to be a stable API).

Separately: given that all of this is still at the "musing" stage, I much prefer going with approach 2 to unblock the specific customer who needs this.

**Comment by umanwizard** [comment]
Having spoken with @ggevay , I am convinced that we should never rely on null inference to predict nullability when expected, and so we should solve this by adding `FORCE NOT NULL` to `FORMAT AVRO`.

**Comment by benesch** [comment]
Sounds good to me.

**Comment by benesch** [comment]
@umanwizard has proposed a third option ([here](https://github.com/MaterializeInc/materialize/issues/17668#issuecomment-1747202062)): add `FORCE NOT NULL` syntax to `CREATE MATERIALIZED VIEW`. This neatly blends the two approaches. It's in line with our philosophy that the materialized view is the "public API" where types and nullability information should live, but a `FORCE NOT NULL` option that is attached to the MV rather than embedded in the `ScalarExpr`

Candidate syntax:

```
-- Trailing WITH options. Perhaps ambiguous?
CREATE MATERIALIZED VIEW ...
WITH (
    FORCE NOT NULL col_a,
    FORCE NOT NULL col_b
)
AS
SELECT ...
```

One other killer feature of this that @umanwizard points out: it sidesteps the immediate need for #17668. There's a plausible story for determining whether your sink has an error: `SELECT * FROM underlying_mv LIMIT 0`. That query will only succeed if there are no errors in the error stream.

**Comment by benesch** [comment]
@umanwizard proposes using `ASSERT` rather than `FORCE`, which I also like better:

```sql
CREATE MATERIALIZED VIEW ...
WITH (
    ASSERT NOT NULL col_a,
    ASSERT NOT NULL col_b
)
AS
SELECT ...
```

**Comment by benesch** [comment]
Fixed by #22248.
